app/models/transaction.rb

```diff
@@ -13,4 +13,15 @@ class Transaction < ApplicationRecord
     :receiving_organisation_type
   validates :value, numericality: {other_than: 0, less_than_or_equal_to: 99_999_999_999.00}
   validates :date, date_not_in_future: true, date_within_boundaries: true
+
+  before_save :set_financial_quarter_from_date
```

Comment:
Did you settle on this hook rather than in the create transaciton service as we expect this all to go away when we collect the financial quarter in the interface?

What is you take on using Rails hooks? Lots of folks tend to avoid them - I am not sure where I land yet?

Reply:
My strategy here is that I want the app to start converting all input dates into quarters implicitly, no matter how the transaction is created, so it will affect the web form, the bulk importer, etc.

Then we deploy this change, which gets the app to start adding a quarter to all new transactions, and back-fills all the existing ones.

Once that behaviour is in place, we change the UI so you input a quarter rather than a date, and make the model convert the quarter into date, rather than the other way around.

I tend not to use hooks often, and usually only to ensure the internal consistency of a single record, not to perform other side effects since the control flow rapidly becomes confusing and causes surprises. For example here I'm making sure two fields agree with each other, I'll often use hooks to "clean up" data before validation e.g. removing leading/trailing space, etc. But not to, say, send an email -- that doesn't belong in the data layer.

Reply:
I also only use hooks when you definitely always want that behaviour to implicitly happen on creating/saving a record. If you ever want to create records in a different state in the app or in tests, then that logic does not belong in a hook -- it's something the caller should be doing explicitly.